go/arrow/decimal256/decimal256.go

```diff
@@ -0,0 +1,569 @@
+	v -= float32(math.Ldexp(float64(arr[2]), 128))
+	arr[1] = float32(math.Floor(math.Ldexp(float64(v), -64)))
+	v -= float32(math.Ldexp(float64(arr[1]), 64))
+	arr[0] = v
+
+	debug.Assert(arr[3] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[3] < 1.8446744e+19, "bad conversion float32 to decimal256") // 2**64
+	debug.Assert(arr[2] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[2] < 1.8446744e+19, "bad conversion float32 to decimal256") // 2**64
+	debug.Assert(arr[1] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[1] < 1.8446744e+19, "bad conversion float32 to decimal256") // 2**64
+	debug.Assert(arr[0] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[0] < 1.8446744e+19, "bad conversion float32 to decimal256") // 2**64
+	return Num{[4]uint64{uint64(arr[0]), uint64(arr[1]), uint64(arr[2]), uint64(arr[3])}}, nil
+}
+
+func FromFloat64(v float64, prec, scale int32) (Num, error) {
+	debug.Assert(prec > 0 && prec <= 76, "invalid precision for converting to decimal256")
+
+	if math.IsInf(v, 0) {
+		return Num{}, fmt.Errorf("cannot convert %f to decimal256", v)
+	}
+
+	if v < 0 {
+		dec, err := fromPositiveFloat64(-v, prec, scale)
+		if err != nil {
+			return dec, err
+		}
+		return dec.Negate(), nil
+	}
+	return fromPositiveFloat64(v, prec, scale)
+}
+
+func fromPositiveFloat64(v float64, prec, scale int32) (Num, error) {
+	var pscale float64
+	if scale >= -76 && scale <= 76 {
+		pscale = float64PowersOfTen[scale+76]
+	} else {
+		pscale = math.Pow10(int(scale))
+	}
+
+	v *= pscale
+	v = math.RoundToEven(v)
+	maxabs := float64PowersOfTen[prec+76]
+	if v <= -maxabs || v >= maxabs {
+		return Num{}, fmt.Errorf("cannot convert %f to decimal256(precision=%d, scale=%d): overflow",
+			v, prec, scale)
+	}
+
+	var arr [4]float64
+	arr[3] = math.Floor(math.Ldexp(v, -192))
+	v -= math.Ldexp(arr[3], 192)
+	arr[2] = math.Floor(math.Ldexp(v, -128))
+	v -= math.Ldexp(arr[2], 128)
+	arr[1] = math.Floor(math.Ldexp(v, -64))
+	v -= math.Ldexp(arr[1], 64)
+	arr[0] = v
+
+	debug.Assert(arr[3] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[3] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[2] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[2] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[1] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[1] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[0] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[0] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	return Num{[4]uint64{uint64(arr[0]), uint64(arr[1]), uint64(arr[2]), uint64(arr[3])}}, nil
+}
+
+func (n Num) tofloat32Positive(scale int32) float32 {
+	const twoTo64 float32 = 1.8446744e+19
+	if n.arr[3] != 0 || n.arr[2] != 0 {
+		return floatInf
```

Comment:
The decimal256 might have a large integer value but with a scale that compensates...